File: src/main/java/com/home/demos/samplebankmodule/repositories/cards/impl/CardRepositoryImpl.java
package com.home.demos.samplebankmodule.repositories.cards.impl;

import com.home.demos.samplebankmodule.repositories.cards.CardRepository;
import com.home.demos.samplebankmodule.repositories.cards.entities.Card;
import org.apache.commons.dbutils.QueryRunner;
import org.apache.commons.dbutils.handlers.BeanListHandler;
import org.apache.commons.dbutils.handlers.ScalarHandler;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Repository;

import java.sql.Connection;
import java.util.List;
import java.util.Objects;

@Repository
public class CardRepositoryImpl implements CardRepository {

    private final QueryRunner queryRunner;

    @Autowired
    public CardRepositoryImpl(QueryRunner queryRunner) {
        this.queryRunner = queryRunner;
    }

    @Override
    public List<Card> findAllByClientId(Long clientId) {

        try (Connection connection = queryRunner.getDataSource().getConnection()) {
            return queryRunner.query(
                    connection,
                    "SELECT * FROM cards WHERE client_id=?",
                    new BeanListHandler<>(Card.class),
                    clientId
            );
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public Card save(Card card) {
        try (Connection connection = queryRunner.getDataSource().getConnection()) {
            Long id = queryRunner.insert(
                    connection,
                    "INSERT INTO cards (title, clientId, sum) VALUES(?,?,?)",
                    new ScalarHandler<>(),
                    card.getTitle(), card.getClientId(), card.getSum()
            );

            card.setId(id);

            return card;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/kotlin/com/home/demos/samplebankmodule/repositories/cards/CardRepositoryImpl.kt
package com.home.demos.samplebankmodule.repositories.cards

import com.home.demos.samplebankmodule.model.Card
import org.apache.commons.dbutils.QueryRunner
import org.apache.commons.dbutils.handlers.BeanListHandler
import org.apache.commons.dbutils.handlers.ScalarHandler

class CardRepositoryImpl(private val queryRunner: QueryRunner) : CardRepository {

    override fun findAllByClientId(clientId: Long): List<Card> {
        return queryRunner.dataSource.connection.use { connection ->
            queryRunner.query(
                    connection,
                    "SELECT * FROM cards WHERE client_id=",
                    BeanListHandler(Card::class.java),
                    clientId
            )
        }
    }

    override fun save(card: Card): Card {
        return queryRunner.dataSource.connection.use { connection ->
            val id = queryRunner.insert(
                    connection,
                    "INSERT INTO cards (title, clientId, sum) VALUES(,,)",
                    ScalarHandler<Long>(),
                    card.title, card.clientId, card.sum
            )
            Card(id, card.title, card.clientId, card.sum);
        }
    }
}


